### DISCUSSION TITLE
.NET Core Api Referencing .NET Framework Assembly fails with ReflectionTypeLoadException

### DISCUSSION THREAD

**Discussion by terryaney** [xunit/xunit#2923] [Question]
I'm just starting the process of migrating an API to .NET Core.  The Api needs to create a fire & forget job via Hangfire.  The issue is that the job/delegate that is used to enqueue the Hangfire job needed to reference a .NET Framework assembly because that is not getting migrated yet.

I added the reference to my .NET Framework assembly and called via this line:

```csharp
var hangfireId = HF.BackgroundJob.Enqueue( () => 
	new BTR.Evolution.Hangfire.Schedulers.JobInvoker().Invoke(
		$"Email Blast From {authId}",
		inputPackage.ToString(),
		null,
		HF.JobCancellationToken.Null 
	)
);
```

I've ran the API locally and submitted to it via Postman and the API successfully schedules that Hangfire job (and it executes) without issue.

However, when I try to run my integration tests against the api (and currently I'm not even testing this method, I just have a skeleton test project enabled for someone else to step in and start building tests), I get the following exception on all 3 of my endpoints that are tested.

I *think* this is something going on with xUnit (or my misuse) vs FastEndpoints since the Api works on its own??

> Starting test execution, please wait...
> A total of 1 test files matched the specified pattern.
> [xUnit.net 00:00:06.19]     Camelot.Api.Excel.Tests.Integration.xDSDataEndpointTests.GetTests.Get_ReturnsValidation_WhenParametersMissing [FAIL]
> [xUnit.net 00:00:06.20]     Camelot.Api.Excel.Tests.Integration.CalcEngineEndpointTests.DownloadDebugEndpointTests.Get_ReturnsValidation_WhenParametersMissing [FAIL]
> [xUnit.net 00:00:06.20]     Camelot.Api.Excel.Tests.Integration.CalcEngineEndpointTests.DebugListingEndpointTests.Get_ReturnsValidation_WhenParametersMissing [FAIL]
>   Failed Camelot.Api.Excel.Tests.Integration.xDSDataEndpointTests.GetTests.Get_ReturnsValidation_WhenParametersMissing [1 ms]
>   Error Message:
>    System.Reflection.ReflectionTypeLoadException : Unable to load one or more of the requested types.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
> Could not load file or assembly 'System.Data.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The system cannot find the file specified.
>   Stack Trace:
>      at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.Assembly.GetTypes()
>    at FastEndpoints.EndpointData.<>c.<BuildEndpointDefinitions>b__8_3(Assembly a)
>    at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
>    at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
>    at FastEndpoints.EndpointData.BuildEndpointDefinitions(EndpointDiscoveryOptions options)
>    at FastEndpoints.EndpointData.<>c__DisplayClass6_0.<.ctor>b__0()
>    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
>    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
>    at System.Lazy`1.CreateValue()
>    at FastEndpoints.EndpointData.get_Found()
>    at FastEndpoints.MainExtensions.MapFastEndpoints(IEndpointRouteBuilder app, Action`1 configAction)
>    at FastEndpoints.MainExtensions.UseFastEndpoints(IApplicationBuilder app, Action`1 configAction)
>    at KAT.Camelot.Api.ConfigurationExtensions.UseApiConfiguration[TLogger](IApplicationBuilder app, Action`1 options) in C:\BTR\Camelot\Core\Api\src\Extensions\ConfigurationExtensions.cs:line 316
>    at Program.<Main>$(String[] args) in C:\BTR\Camelot\Api\Excel\src\Program.cs:line 47
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
>    at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
> --- End of stack trace from previous location ---
>    at Microsoft.AspNetCore.Mvc.Testing.DeferredHostBuilder.DeferredHost.StartAsync(CancellationToken cancellationToken)
>    at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Start(IHost host)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateHost(IHostBuilder builder)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.ConfigureHostBuilder(IHostBuilder hostBuilder)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.EnsureServer()
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(DelegatingHandler[] handlers)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(Uri baseAddress, DelegatingHandler[] handlers)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient(WebApplicationFactoryClientOptions options)
>    at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient()
>    at KAT.Camelot.Testing.Integration.BaseFactory`1.CreateClient(Func`1 getJwtInfo, Dictionary`2 claims) in C:\BTR\Camelot\Core\Testing\Integration\src\[Core]\BaseFactory.cs:line 114
>    at KAT.Camelot.Testing.Integration.BaseFactory`1.CreateClient(Func`1 getJwtInfo) in C:\BTR\Camelot\Core\Testing\Integration\src\[Core]\BaseFactory.cs:line 74
>    at Camelot.Api.Excel.Tests.Integration.ApiFactory.CreateClient() in C:\BTR\Camelot\Api\Excel\tests\Integration\ApiFactory.cs:line 12        
>    at Camelot.Api.Excel.Tests.Integration.xDSDataEndpointTests.GetTests..ctor(ApiFactory factory) in C:\BTR\Camelot\Api\Excel\tests\Integration\xDSDataEndpointTests\GetTests.cs:line 19
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
>    at System.Reflection.ConstructorInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
>   Failed Camelot.Api.Excel.Tests.Integration.CalcEngineEndpointTests.DownloadDebugEndpointTests.Get_ReturnsValidation_WhenParametersMissing [1 ms]

**Comment by bradwilson** [comment]
We don't set up the execution environment or the dependency resolution at all when running in .NET Core. That's done by VSTest.

One way to verify this would be to port a single test to NUnit or MSTest and see if you get the same kind of result via `dotnet test` with them.

**Reply by terryaney** [reply]
Want me to do this verification for your information or you already know result and just for my proof? :)

**Reply by bradwilson** [reply]
I don't actually know the result. This is the first time it's come up, honestly, so I would like to know if there's something we're doing wrong/insufficiently that I could fix.

**Reply by bradwilson** [reply]
Based on what's missing, it looks like the assembly resolution is not considering the GAC (which would make sense, since .NET Core doesn't use the GAC). I don't know what's different about what happens when you're just running the application.

**Reply by terryaney** [reply]
Appears to maybe be an issue with FastEndpoints.  Or rather, how I'm using them when configuring/creating the 'web host' during integration test setup (please excuse my butchering of correct terminology for web host/application terms).  I've asked them for any guidance as well while I dig more.

**Reply by terryaney** [reply]
FYI, I have the following code in my Integration Test fixture...

```csharp
builder.ConfigureTestServices( services =>
{
	// Recommended by Chapsas...doubt any present in this project, but just want here for reference
	services.RemoveAll( typeof( IHostedService ) );

	// few other 'safe' lines....
});
```

Don't really recall why Nick Chapsas recommended this, but it was from one of his Zero to Hero training videos that I watched about Integration Testing.

Is that line suspect at all?

**Reply by terryaney** [reply]
For completeness and to make sure I'm doing things correctly.

1. Removing the `RemoveAll` line above had no effect.  Same error.
2. My XUnit test class converted to MSTest is below and it produces same error (now that I try to pass in the ApiFactory (fixture) in on the ctor, without DI, MSTest worked).

```csharp
public class GetTests : IClassFixture<ApiFactory>
{
	private readonly ApiFactory factory;
	public GetTests( ApiFactory factory )
	{
		this.factory = factory;
	}

	[Fact]
	public async Task Get_ReturnsValidation_WhenParametersMissing()
	{
		// Arrange
		await Task.Delay( 1 );

		// Act

		// Assert 
	}

}

[TestClass]
public class GetMSTests
{
	private readonly ApiFactory factory;
	public GetMSTests( ApiFactory factory )
	{
		this.factory = factory;
	}

	[TestMethod]
	public async Task Get_ReturnsValidation_WhenParametersMissing_MSTest()
	{
		// Arrange
		await Task.Delay(1);

		// Act

		// Assert 
	}
}
```

**Reply by bradwilson** [reply]
> Is that line suspect at all?

Believe it not (given my personal history), I have honestly zero experience testing ASP.NET Core apps in xUnit.net. 😞

**Reply by bradwilson** [reply]
FastEndpoints definitely recommends us for integration testing: https://fast-endpoints.com/docs/integration-unit-testing#integration-testing

I was able to follow their two step getting started:

```
dotnet new install FastEndpoints.TemplatePack
dotnet new feproj -n E2EWalkthrough
```

And then their tests ran:

```
pwsh> dotnet test

  Determining projects to restore...
  Restored C:\Dev\repro\E2EWalkthrough\Source\E2EWalkthrough.csproj (in 2.13 sec).
  Restored C:\Dev\repro\E2EWalkthrough\Tests\Tests.csproj (in 2.13 sec).
  E2EWalkthrough -> C:\Dev\repro\E2EWalkthrough\Source\bin\Debug\net8.0\E2EWalkthrough.dll
  Tests -> C:\Dev\repro\E2EWalkthrough\Tests\bin\Debug\net8.0\Tests.dll
Test run for C:\Dev\repro\E2EWalkthrough\Tests\bin\Debug\net8.0\Tests.dll (.NETCoreApp,Version=v8.0)
Microsoft (R) Test Execution Command Line Tool Version 17.9.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     2, Skipped:     0, Total:     2, Duration: 4 ms - Tests.dll (net8.0)
```

They are using their own customizations around fixtures, so this is their sample test class:

```csharp
using SayHello;

namespace Tests.SayHello;

public class Tests(App App) : TestBase<App>
{
    [Fact, Priority(1)]
    public async Task Invalid_User_Input()
    {
        var (rsp, res) = await App.Client.POSTAsync<Endpoint, Request, ErrorResponse>(
                             new()
                             {
                                 FirstName = "x",
                                 LastName = "y"
                             });

        rsp.StatusCode.Should().Be(HttpStatusCode.BadRequest);
        res.Errors.Count.Should().Be(2);
        res.Errors.Keys.Should().Equal("firstName", "lastName");
    }

    [Fact, Priority(2)]
    public async Task Valid_User_Input()
    {
        var (rsp, res) = await App.Client.POSTAsync<Endpoint, Request, Response>(
                             new()
                             {
                                 FirstName = "Mike",
                                 LastName = "Kelso"
                             });

        rsp.StatusCode.Should().Be(HttpStatusCode.OK);
        res.Message.Should().Be("Hello Mike Kelso...");
    }
}
```

and this is App:

```csharp
using Microsoft.AspNetCore.Hosting;

namespace Tests.SayHello;

public class App : AppFixture<Program>
{
    protected override Task SetupAsync()
    {
        // place one-time setup for the fixture here
        return Task.CompletedTask;
    }

    protected override void ConfigureApp(IWebHostBuilder a)
    {
        // do host builder configuration here
    }

    protected override void ConfigureServices(IServiceCollection s)
    {
        // do test service registration here
    }

    protected override Task TearDownAsync()
    {
        // do cleanups here
        return Task.CompletedTask;
    }
}
```

Maybe that can send you in the right direction...?

**Reply by terryaney** [reply]
Thanks a lot for peeking into it.  I actually have no problem running FastEndpoints/XUnit together in other apis (I should have stated that).  The problem only presented itself when I referenced .NET Framework.  They asked for a sample repository and if you wanted to peek, you can see it [here](https://github.com/terryaney/FastEndpoints.Repro/tree/main/CoreRefFramework).  I'll update this question if/when I resolve it.  

I'll have to see what they have going for fixture customization, sounds interesting.
